akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -380,3 +361,150 @@ private[akka] final class SinkQueueAdapter[T](delegate: SinkQueue[T]) extends ak
   import akka.dispatch.ExecutionContexts.{ sameThreadExecutionContext ⇒ same }
   def pull(): CompletionStage[Optional[T]] = delegate.pull().map(_.asJava)(same).toJava
 }
+
+private[stream] object AdvancedPublisherSink {
+  sealed trait PublisherSinkMessages[+T]
+  private case class RequestMore[T](subscriber: SubscriptionWithCursor[T], demand: Long) extends PublisherSinkMessages[T]
+  private case class Cancel[T](subscriber: SubscriptionWithCursor[T]) extends PublisherSinkMessages[T]
+  private case class Subscribe[T](subscriber: Subscriber[_ >: T]) extends PublisherSinkMessages[T]
+
+  sealed trait Directive
+  case object Fanout extends Directive
+  case object FanoutWithDrainIfNoSubscribtions extends Directive
+  case object DrainIfNoSubscribtions extends Directive
+
+  def apply[T](fanout: Boolean, finalizeOnLastSubscriptionCompletion: Boolean): AdvancedPublisherSink[T] =
+    new AdvancedPublisherSink[T](getDirective(fanout, finalizeOnLastSubscriptionCompletion))
+
+  def getDirective(fanout: Boolean, finalizeOnLastSubscriptionCompletion: Boolean): Directive = {
+    (fanout, finalizeOnLastSubscriptionCompletion) match {
+      case (true, true)   ⇒ Fanout
+      case (true, false)  ⇒ FanoutWithDrainIfNoSubscribtions
+      case (false, false) ⇒ DrainIfNoSubscribtions
+      case _              ⇒ throw new IllegalArgumentException("For single subscription use Publisher Sink")
+    }
```

Comment:
I think if feels a bit over-abstracted to transform these two booleans to an enumeration of strategies, when I look at the stage how they are used I think it would make more sense if the comparisons were `if (fanout) ...` or `if (finalizeOnLastSubscriptionCompletion) ...`.